app/src/ui/diff/text-diff.tsx

```diff
@@ -1381,10 +1390,88 @@ export class TextDiff extends React.Component<ITextDiffProps, ITextDiffState> {
 
   public componentWillUnmount() {
     this.cancelSelection()
+    this.unmountWhitespaceHint()
     this.codeMirror = null
     document.removeEventListener('find-text', this.onFindText)
   }
 
+  private mountWhitespaceHint(index: number) {
+    this.unmountWhitespaceHint()
+
+    // Since we're in a bit of a weird state here where CodeMirror is mounted
+    // through React and we're in turn mounting a React component from a
+    // DOM event we want to make sure we're not mounting the Popover
+    // synchronously. Doing so will cause the popover to receiving the bubbling
+    // mousedown event (on document) which caused it to be mounted in the first
+    // place and it will then close itself thinking that it's seen a mousedown
+    // event outside of its container.
+    this.whitespaceHintMountId = requestAnimationFrame(() => {
+      this.whitespaceHintMountId = null
+      const cm = this.codeMirror
+
+      if (cm === null) {
+        return
+      }
+
+      const container = document.createElement('div')
+      container.style.position = 'absolute'
+      const scroller = cm.getScrollerElement()
+
+      const diffSize = getLineWidthFromDigitCount(
+        getNumberOfDigits(this.state.diff.maxLineNumber)
+      )
+
+      const lineY = cm.heightAtLine(index, 'local')
+      // We're positioning relative to the scroll container, not the
+      // sizer or lines so we'll have to account for the gutter width and
+      // the hunk handle.
+      const style: React.CSSProperties = { left: diffSize * 2 + 10 }
+      let caretPosition = PopoverCaretPosition.LeftTop
+
+      // Offset down by 10px to align the popover arrow.
+      container.style.top = `${lineY - 10}px`
+
+      // If the line is further than 50% down the viewport we'll flip the
+      // popover to point upwards so as to not get hidden beneath (or above)
+      // the scroll boundary.
+      if (lineY - scroller.scrollTop > scroller.clientHeight / 2) {
+        caretPosition = PopoverCaretPosition.LeftBottom
+        style.bottom = -35
+      }
```

Comment:
This is a nice touch! I see it's not in the side-by-side diff, which instead pushes the scroll down:
![Kapture 2021-09-22 at 17 37 33](https://user-images.githubusercontent.com/1083228/134375405-ef035ebf-c66f-4431-9083-2d9386ecf6f7.gif)

I don't think we _need_ to add complexity for such an edge case but just wanted to let you know!

Reply:
Yeah, I looked into this and I'll reply to both your and @tidy-dev's comments on this here. Due to the way the `List` component works (hiding overflow) we can't easily have the popover extend above the diff so there would always be scenarios where we'd extend the height of the diff (single line/smalls diffs) so I figured we'll live with this annoyance for now and see if we can't fix it by means of a more flexible popover system in the future.